src/window.rs

```diff
@@ -0,0 +1,138 @@
+use core::marker::PhantomData;
+use core::f64;
+
+use {Sample, FloatSample};
+use signal;
+use signal::{Phase, ConstHz, Step};
+use conv::{FromSample, ToSample};
+use frame::Frame;
+
+// Using types instead of enum allows for static dispatch
+pub trait WindowType<S: Sample, T: Step> {
+    fn at_phase(phase: S) -> S;
+}
+
+pub struct HanningWindow;
+
+impl<S: Sample + ToSample<f64> + FromSample<f64>, T: Step> WindowType<S, T> for HanningWindow {
+    fn at_phase(phase: S) -> S {
+        let v = phase.to_sample::<f64>() * f64::consts::PI * 2.;
+        (0.5f64 * (1f64 - v.cos())).to_sample::<S>()
+    }
+}
+
+pub struct RectangleWindow;
+
+impl<S: Sample + FromSample<f64>, T: Step> WindowType<S, T> for RectangleWindow {
+    #[allow(unused)]
+    fn at_phase(phase: S) -> S {
+        (1.).to_sample::<S>()
+    }
+}
+
+pub struct Window<S: Sample, T: Step, F: Frame<Sample=S>, WT: WindowType<S, T>> {
```

Comment:
Unfortunately it seems like we're stuck here. The occasional `FromSample<f64>` bound on `S` means that I can't (to my knowledge) only use an associated type here. Could be something I don't know though—getting close to trait arcana here.